node/core/av-store/src/lib.rs

```diff
@@ -626,7 +647,7 @@ where
 async fn request_candidate_events<Context>(
 	ctx: &mut Context,
 	hash: Hash,
-) -> Result<Vec<CandidateEvent>, Error>
+) -> Result<Vec<CandidateEvent>, NonFatalError>
```

Comment:
If we know that all that is returned here, we should probably let the caller decide on what to do with the error rather than handling most of them within this `fn` already?

Reply:
The caller is `process_block_activated` in this subsystem, it's handling the error. Not sure what exactly you're suggesting?

Reply:
In the following lines quite a few channel send error cases are silently dropped. If we know that all this fn returns is `NonFatalError`s anyways, we can also return those and let the caller handle the printing. It's a nit :) - generally I just prefer to have nicely testable `fn`s with well defined bounds (here it gets fuzzy with what non-fatal error is logged and which is returned).

Reply:
In case of `process_message`, it actually returns hard errors (db or `get_chunk` failures), but silences `NonFatalError`s.
Are you suggesting we return something like `Result<Result<_, Error>, NonFatalError>` instead? It would lose the context that we have in `process_message` when logging errors.

Reply:
An alternative would be something like

```rust
trait PolkadotError {
  fn is_fatal(&self) -> bool;
}
```

which, combined with requiring subsystems to return a `PolkadotError`, would mean that we could get away with just a single error type per subsystem, and let the overseer log and ignore all non-fatal errors.

Reply:
Downside would be to either require a macro to `#[soft]` (or so) to some error variants or impl the trait manually all over the place (juck).

Reply:
I don't think that a macro which lets us tag certain variants as `#[soft]` would be that bad to use in practice, and it would be good fun to write the macro.

The main problem with this idea IMO is that all errors are hard in constructors, even those tagged as soft, but I don't see that as being a huge deal.

Reply:
Every subsystem has to differentiate between the fatal and non-fatal error in order to decide when to shutdown in `fn run`, so they might as well log the non-fatal errors. 

I think it would be nice to utilize the type system somehow to ensure we can't accidentally shutdown on `NonFatalError` error, e.g. by not implementing `std::error:Error` or something.